mqtt/mqtt-bridge/src/persist/waking_state/memory.rs

```diff
@@ -0,0 +1,58 @@
+#![allow(dead_code)] // TODO remove when ready
+
+use std::{
+    cmp::min,
+    collections::{HashMap, VecDeque},
+    task::Waker,
+};
+
+use mqtt3::proto::Publication;
+
+use crate::persist::{waking_state::StreamWakeableState, Key, PersistError};
+
+/// When elements are retrieved they are moved to the in flight collection.
+pub struct WakingMemoryStore {
+    queue: VecDeque<(Key, Publication)>,
+    in_flight: HashMap<Key, Publication>,
+    waker: Option<Waker>,
+}
+
+impl WakingMemoryStore {
+    pub fn new() -> Self {
+        WakingMemoryStore {
+            queue: VecDeque::new(),
+            in_flight: HashMap::new(),
+            waker: None,
+        }
+    }
+}
+
+impl StreamWakeableState for WakingMemoryStore {
+    fn insert(&mut self, key: Key, value: Publication) -> Result<(), PersistError> {
+        self.queue.push_back((key, value));
+
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+
+        Ok(())
+    }
+
+    fn batch(&mut self, count: usize) -> Result<VecDeque<(Key, Publication)>, PersistError> {
+        let count = min(count, self.queue.len());
+        let output: VecDeque<_> = self.queue.drain(..count).collect();
```

Comment:
I just looked at the docs for drain and see that it can panic. Although it should never do this because we guarantee count is at most the length of the queue, it makes me scared. Should we change this approach?

Reply:
Any alternatives? for me this is fine as long as we guarantee the count is at most the length of the queue